File: _includes/code/csharp/ManageCollectionsMigrateDataTest.cs
using Xunit;
using Weaviate.Client;
using Weaviate.Client.Models;
using System;
using System.Threading.Tasks;
using System.Linq;
using System.Collections.Generic;
using System.Text.Json;

namespace WeaviateProject.Tests;

public class ManageCollectionsMigrateDataTest : IAsyncLifetime
{
    private static WeaviateClient clientSrc;
    private static WeaviateClient clientTgt;
    private const int DATASET_SIZE = 50;

    // Defines the schema structure for strong typing
    private class WineReviewModel
    {
        public string title { get; set; }
        public string review_body { get; set; }
        public string country { get; set; }
        public int? points { get; set; }
        public double? price { get; set; }
    }

    public async Task InitializeAsync()
    {
        clientSrc = await Connect.Local(restPort: 8080, grpcPort: 50051);
        clientTgt = await Connect.Local(restPort: 8090, grpcPort: 50061);

        // Ensure clean state for source collections
        if (await clientSrc.Collections.Exists("WineReview"))
        {
            await clientSrc.Collections.Delete("WineReview");
        }
        if (await clientSrc.Collections.Exists("WineReviewMT"))
        {
            await clientSrc.Collections.Delete("WineReviewMT");
        }

        await CreateCollection(clientSrc, "WineReview", false);
        await CreateCollection(clientSrc, "WineReviewMT", true);

        var wineReview = clientSrc.Collections.Use("WineReview");
        var wineReviewData = Enumerable.Range(0, DATASET_SIZE)
            .Select(i => new WineReviewModel
            {
                title = $"Review {i}",
                review_body = "Description...",
                price = 10.5 + i
            })
            .ToArray();

        await wineReview.Data.InsertMany(wineReviewData);

        var wineReviewMT = clientSrc.Collections.Use("WineReviewMT");
        await wineReviewMT.Tenants.Add(["tenantA"]);
        await wineReviewMT.WithTenant("tenantA").Data.InsertMany(wineReviewData);
    }

    public async Task DisposeAsync()
    {
        await clientSrc.Collections.DeleteAll();
        await clientTgt.Collections.DeleteAll();
    }

    // START CreateCollectionCollectionToCollection // START CreateCollectionCollectionToTenant // START CreateCollectionTenantToCollection // START CreateCollectionTenantToTenant
    private static async Task<CollectionClient> CreateCollection(WeaviateClient clientIn,
        string collectionName, bool enableMt)
    {
        // END CreateCollectionCollectionToCollection // END CreateCollectionCollectionToTenant // END CreateCollectionTenantToCollection // END CreateCollectionTenantToTenant
        if (await clientIn.Collections.Exists(collectionName))
        {
            await clientIn.Collections.Delete(collectionName);
        }
        // START CreateCollectionCollectionToCollection // START CreateCollectionCollectionToTenant // START CreateCollectionTenantToCollection // START CreateCollectionTenantToTenant
        return await clientIn.Collections.Create(new CollectionConfig
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = enableMt },
            VectorConfig = new VectorConfig("default", new Vectorizer.Text2VecTransformers()),
            Properties =
            [
                Property.Text("review_body"),
                Property.Text("title"),
                Property.Text("country"),
                Property.Int("points"),
                Property.Number("price")
            ]
        });
    }
    // END CreateCollectionCollectionToCollection // END CreateCollectionCollectionToTenant // END CreateCollectionTenantToCollection // END CreateCollectionTenantToTenant
    // TODO[g-despot] NEW: Why can't I insert many with preserved IDs?

    // START CollectionToCollection // START TenantToCollection // START CollectionToTenant // START TenantToTenant
    private async Task MigrateData<T>(CollectionClient collectionSrc,
     CollectionClient collectionTgt)
    {
        Console.WriteLine("Starting data migration...");

        // Fetch source objects
        var response = await collectionSrc.Query.FetchObjects(limit: 10000, includeVectors: true);

        // Map to Strong Type List
        var sourceObjects = new List<T>();
        foreach (var obj in response.Objects)
        {
            // Deserialize the inner properties Dictionary to the POCO type
            var json = JsonSerializer.Serialize(obj.Properties);
            var typedObj = JsonSerializer.Deserialize<T>(json);
            if (typedObj != null)
            {
                sourceObjects.Add(typedObj);
            }
        }

        // InsertMany using Strong Types
        await collectionTgt.Data.InsertMany(sourceObjects.ToArray());

        Console.WriteLine($"Data migration complete. Migrated {sourceObjects.Count} objects.");
    }
    // END CollectionToCollection // END TenantToCollection // END CollectionToTenant // END TenantToTenant

    private async Task<bool> VerifyMigration(CollectionClient collectionTgt, int expectedCount)
    {
        // Verification modified because InsertMany generates NEW IDs.
        // We check if the total count matches and if a sample query works.
        var countResult = await collectionTgt.Aggregate.OverAll(totalCount: true);

        if (countResult.TotalCount != expectedCount)
        {
            Console.WriteLine($"Count mismatch. Expected {expectedCount}, found {countResult.TotalCount}");
            return false;
        }

        var sample = await collectionTgt.Query.FetchObjects(limit: 1);
        if (sample.Objects.Count == 0 || !sample.FirstOrDefault().Properties.ContainsKey("title"))
        {
            Console.WriteLine("Data verification failed. Properties missing.");
            return false;
        }

        Console.WriteLine("Verification successful!");
        return true;
    }

    // START CreateCollectionCollectionToCollection
    private async Task CreateCollectionToCollection()
    {
        await CreateCollection(clientTgt, "WineReview", false);
    }
    // END CreateCollectionCollectionToCollection

    [Fact]
    // START CollectionToCollection
    public async Task TestCollectionToCollection()
    {
        await CreateCollectionToCollection();

        var reviewsSrc = clientSrc.Collections.Use("WineReview");
        var reviewsTgt = clientTgt.Collections.Use("WineReview");

        // Pass the Type to the generic method
        // END CollectionToCollection
        await MigrateData<WineReviewModel>(reviewsSrc, reviewsTgt);

        Assert.True(await VerifyMigration(reviewsTgt, DATASET_SIZE));
        // START CollectionToCollection
    }
    // END CollectionToCollection

    // START CreateCollectionTenantToCollection
    private async Task CreateTenantToCollection()
    {
        await CreateCollection(clientTgt, "WineReview", false);
    }
    // END CreateCollectionTenantToCollection

    [Fact]
    // START TenantToCollection
    public async Task TestTenantToCollection()
    {
        await CreateTenantToCollection();

        var reviewsSrc = clientSrc.Collections.Use("WineReviewMT");
        var reviewsTgt = clientTgt.Collections.Use("WineReview");
        var reviewsSrcTenantA = reviewsSrc.WithTenant("tenantA");

        await MigrateData<WineReviewModel>(reviewsSrcTenantA, reviewsTgt);

        // END TenantToCollection
        Assert.True(await VerifyMigration(reviewsTgt, DATASET_SIZE));
        // START TenantToCollection
    }
    // END TenantToCollection

    // START CreateCollectionCollectionToTenant
    private async Task CreateCollectionToTenant()
    {
        await CreateCollection(clientTgt, "WineReviewMT", true);
    }
    // END CreateCollectionCollectionToTenant

    // START CreateTenants // START CreateCollectionTenantToTenant
    private async Task CreateTenants()
    {
        var reviewsMtTgt = clientTgt.Collections.Use("WineReviewMT");

        var tenantsTgt = new[] { new Tenant { Name = "tenantA" }, new Tenant { Name = "tenantB" } };
        await reviewsMtTgt.Tenants.Add(tenantsTgt);
    }
    // END CreateTenants // END CreateCollectionTenantToTenant

    [Fact]
    // START CollectionToTenant
    public async Task TestCollectionToTenant()
    {
        await CreateCollectionToTenant();
        await CreateTenants();

        var reviewsMtTgt = clientTgt.Collections.Use("WineReviewMT");
        var reviewsSrc = clientSrc.Collections.Use("WineReview");

        var reviewsTgtTenantA = reviewsMtTgt.WithTenant("tenantA");

        await MigrateData<WineReviewModel>(reviewsSrc, reviewsTgtTenantA);
        // END CollectionToTenant

        Assert.True(await VerifyMigration(reviewsTgtTenantA, DATASET_SIZE));
        // START CollectionToTenant
    }
    // END CollectionToTenant

    // START CreateCollectionTenantToTenant
    private async Task CreateTenantToTenant()
    {
        await CreateCollection(clientTgt, "WineReviewMT", true);
    }
    // END CreateCollectionTenantToTenant

    [Fact]
    // START TenantToTenant
    public async Task TestTenantToTenant()
    {
        await CreateTenantToTenant();
        await CreateTenants();

        var reviewsMtSrc = clientSrc.Collections.Use("WineReviewMT");
        var reviewsMtTgt = clientTgt.Collections.Use("WineReviewMT");
        var reviewsSrcTenantA = reviewsMtSrc.WithTenant("tenantA");
        var reviewsTgtTenantA = reviewsMtTgt.WithTenant("tenantA");

        await MigrateData<WineReviewModel>(reviewsSrcTenantA, reviewsTgtTenantA);
        // END TenantToTenant

        Assert.True(await VerifyMigration(reviewsTgtTenantA, DATASET_SIZE));
        // START TenantToTenant
    }
    // END TenantToTenant
}

File: _includes/code/java-v6/src/test/java/ManageCollectionsMigrateDataTest.java
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.collections.CollectionHandle;
import io.weaviate.client6.v1.api.collections.Generative;
import io.weaviate.client6.v1.api.collections.Property;
import io.weaviate.client6.v1.api.collections.VectorConfig;
import io.weaviate.client6.v1.api.collections.WeaviateObject;
import io.weaviate.client6.v1.api.collections.tenants.Tenant;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import static org.assertj.core.api.Assertions.assertThat;

class ManageCollectionsMigrateDataTest {

  private static WeaviateClient clientSrc;
  private static WeaviateClient clientTgt;
  private static final int DATASET_SIZE = 50;

  @BeforeAll
  public static void beforeAll() throws IOException {
    String openaiApiKey = System.getenv("OPENAI_API_KEY");
    // Connect to the source Weaviate instance
    clientSrc = WeaviateClient.connectToLocal(
        config -> config.setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));
    // Connect to the target Weaviate instance
    clientTgt = WeaviateClient.connectToLocal(config -> config.port(8090)
        .grpcPort(50061)
        .setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));


    // Simulate weaviate-datasets by creating and populating source collections
    createCollection(clientSrc, "WineReview", false);
    createCollection(clientSrc, "WineReviewMT", true);

    var wineReview = clientSrc.collections.use("WineReview");
    List<Map<String, Object>> wineReviewData = new ArrayList<>();
    for (int i = 0; i < DATASET_SIZE; i++) {
      wineReviewData.add(Map.of("title", "Review " + i));
    }
    wineReview.data.insertMany(wineReviewData.toArray(new Map[0]));

    var wineReviewMT = clientSrc.collections.use("WineReviewMT");
    wineReviewMT.tenants.create(List.of(Tenant.active("tenantA")));
    wineReviewMT.withTenant("tenantA").data
        .insertMany(wineReviewData.toArray(new Map[0]));

    assertThat(clientSrc.isReady()).isTrue();
  }

  @AfterAll
  public static void afterAll() throws Exception {
    if (clientSrc != null) {
      clientSrc.close();
    }
    if (clientTgt != null) {
      clientTgt.close();
    }
  }

  @AfterEach
  public void cleanupTarget() throws IOException {
    // Clean up collections on the target client after each test
    clientTgt.collections.delete("WineReview");
    clientTgt.collections.delete("WineReviewMT");
  }

  // START CreateCollectionCollectionToCollection // START CreateCollectionCollectionToTenant // START CreateCollectionTenantToCollection // START CreateCollectionTenantToTenant
  private static CollectionHandle<Map<String, Object>> createCollection(
      WeaviateClient clientIn, String collectionName, boolean enableMt)
      throws IOException {
    // END CreateCollectionCollectionToCollection // END CreateCollectionCollectionToTenant // END CreateCollectionTenantToCollection // END CreateCollectionTenantToTenant
    if (clientIn.collections.exists(collectionName)) {
      clientIn.collections.delete(collectionName);
    }
    // START CreateCollectionCollectionToCollection // START CreateCollectionCollectionToTenant // START CreateCollectionTenantToCollection // START CreateCollectionTenantToTenant
    clientIn.collections.create(collectionName,
        col -> col.multiTenancy(c -> c.enabled(enableMt))
            // Additional settings not shown
            .vectorConfig(VectorConfig.text2vecTransformers())
            .generativeModule(Generative.cohere())
            .properties(Property.text("review_body"), Property.text("title"),
                Property.text("country"), Property.integer("points"),
                Property.number("price")));

    return clientIn.collections.use(collectionName);
  }

  // END CreateCollectionCollectionToCollection // END CreateCollectionCollectionToTenant // END CreateCollectionTenantToCollection // END CreateCollectionTenantToTenant

  // START CollectionToCollection // START TenantToCollection // START CollectionToTenant // START TenantToTenant
  private void migrateData(CollectionHandle<Map<String, Object>> collectionSrc,
      CollectionHandle<Map<String, Object>> collectionTgt) {
    System.out.println("Starting data migration...");
    List<WeaviateObject<Map<String, Object>>> sourceObjects = StreamSupport
        .stream(collectionSrc.paginate(p -> p.includeVector()).spliterator(),
            false)
        .map(readObj -> WeaviateObject
            .<Map<String, Object>>of(c -> c.properties(readObj.properties())
                .uuid(readObj.uuid())
                .vectors(readObj.vectors())))
        .collect(Collectors.toList());

    collectionTgt.data.insertMany(sourceObjects);

    System.out.println("Data migration complete.");
  }

  // END CollectionToCollection // END TenantToCollection // END CollectionToTenant // END TenantToTenant

  private boolean verifyMigration(
      CollectionHandle<Map<String, Object>> collectionSrc,
      CollectionHandle<Map<String, Object>> collectionTgt, int numSamples) {

    List<WeaviateObject<Map<String, Object>>> srcObjects =
        StreamSupport.stream(collectionSrc.paginate().spliterator(), false)
            .collect(Collectors.toList());

    if (srcObjects.isEmpty()) {
      System.out.println("No objects in source collection");
      return false;
    }
    Collections.shuffle(srcObjects);
    List<WeaviateObject<Map<String, Object>>> sampledObjects =
        srcObjects.subList(0, Math.min(numSamples, srcObjects.size()));

    System.out.printf("Verifying %d random objects...\n",
        sampledObjects.size());
    for (var srcObj : sampledObjects) {
      Optional<WeaviateObject<Map<String, Object>>> tgtObjOpt =
          collectionTgt.query.fetchObjectById(srcObj.uuid());
      if (tgtObjOpt.isEmpty()) {
        System.out.printf("Object %s not found in target collection\n",
            srcObj.uuid());
        return false;
      }
      if (!srcObj.properties().equals(tgtObjOpt.get().properties())) {
        System.out.printf("Properties mismatch for object %s\n", srcObj.uuid());
        return false;
      }
    }
    System.out.println("All sampled objects verified successfully!");
    return true;
  }

  // START CreateCollectionCollectionToCollection
  void createCollectionToCollection() throws IOException {
    createCollection(clientTgt, "WineReview", false);
  }
  // END CreateCollectionCollectionToCollection

  @Test
  // START CollectionToCollection
  void testCollectionToCollection() throws IOException {
    createCollectionToCollection();

    var reviewsSrc = clientSrc.collections.use("WineReview");
    var reviewsTgt = clientTgt.collections.use("WineReview");
    migrateData(reviewsSrc, reviewsTgt);

    // END CollectionToCollection
    assertThat(reviewsTgt.aggregate.overAll(a -> a.includeTotalCount(true))
        .totalCount()).isEqualTo(DATASET_SIZE);
    assertThat(verifyMigration(reviewsSrc, reviewsTgt, 5)).isTrue();
    // START CollectionToCollection
  }
  // END CollectionToCollection

  // START CreateCollectionTenantToCollection
  void createTenantToCollection() throws IOException {
    createCollection(clientTgt, "WineReview", false);
  }
  // END CreateCollectionTenantToCollection

  @Test
  // START TenantToCollection
  void testTenantToCollection() throws IOException {
    createTenantToCollection();

    var reviewsSrc = clientSrc.collections.use("WineReviewMT");
    var reviewsTgt = clientTgt.collections.use("WineReview");
    var reviewsSrcTenantA = reviewsSrc.withTenant("tenantA");
    migrateData(reviewsSrcTenantA, reviewsTgt);

    // END TenantToCollection
    assertThat(reviewsTgt.aggregate.overAll(a -> a.includeTotalCount(true))
        .totalCount()).isEqualTo(DATASET_SIZE);
    assertThat(verifyMigration(reviewsSrcTenantA, reviewsTgt, 5)).isTrue();
    // START TenantToCollection
  }
  // END TenantToCollection

  // START CreateCollectionCollectionToTenant
  void createCollectionToTenant() throws IOException {
    createCollection(clientTgt, "WineReviewMT", true);
  }
  // END CreateCollectionCollectionToTenant

  // START CreateTenants // START CreateCollectionTenantToTenant
  void createTenants() throws IOException {
    var reviewsMtTgt = clientTgt.collections.use("WineReviewMT");

    var tenantsTgt =
        List.of(Tenant.active("tenantA"), Tenant.active("tenantB"));
    reviewsMtTgt.tenants.create(tenantsTgt);
  }
  // END CreateTenants // END CreateCollectionTenantToTenant

  @Test
  // START CollectionToTenant
  void testCollectionToTenant() throws IOException {
    createCollectionToTenant();
    createTenants();

    var reviewsMtTgt = clientTgt.collections.use("WineReviewMT");
    var reviewsSrc = clientSrc.collections.use("WineReview");

    var reviewsTgtTenantA = reviewsMtTgt.withTenant("tenantA");

    migrateData(reviewsSrc, reviewsTgtTenantA);
    // END CollectionToTenant

    assertThat(
        reviewsTgtTenantA.aggregate.overAll(a -> a.includeTotalCount(true))
            .totalCount()).isEqualTo(DATASET_SIZE);
    assertThat(verifyMigration(reviewsSrc, reviewsTgtTenantA, 5)).isTrue();
    // START CollectionToTenant
  }
  // END CollectionToTenant

  // START CreateCollectionTenantToTenant
  void createTenantToTenant() throws IOException {
    createCollection(clientTgt, "WineReviewMT", true);
  }
  // END CreateCollectionTenantToTenant

  @Test
  // START TenantToTenant
  void testTenantToTenant() throws IOException {
    createTenantToTenant();
    createTenants();

    var reviewsMtSrc = clientSrc.collections.use("WineReviewMT");
    var reviewsMtTgt = clientTgt.collections.use("WineReviewMT");
    var reviewsSrcTenantA = reviewsMtSrc.withTenant("tenantA");
    var reviewsTgtTenantA = reviewsMtTgt.withTenant("tenantA");

    migrateData(reviewsSrcTenantA, reviewsTgtTenantA);
    // END TenantToTenant

    assertThat(
        reviewsTgtTenantA.aggregate.overAll(a -> a.includeTotalCount(true))
            .totalCount()).isEqualTo(DATASET_SIZE);
    assertThat(verifyMigration(reviewsSrcTenantA, reviewsTgtTenantA, 5))
        .isTrue();
    // START TenantToTenant

  }
  // END TenantToTenant

}
